### ISSUE TITLE
Super slow open+resize+save on big Jpeg with Mono

### ISSUE THREAD

**Issue by nah0y** [JimBobSquarePants/ImageSharp#94]
### Description
I'm using Mono and Xamarin Studio for that. I'm simply trying to resize a big picture (5184x3456 - 8.2MB) and it takes 12sc to complete.

### Steps to Reproduce
I created a simple console project, added Nuget dependencies to ImageSharp. And I'm using this in the Program.cs:

```
string filePath = ""; // Path to my image

Stopwatch sw1 = new Stopwatch();
sw1.Start();
string tmpFilePath = Path.GetTempPath() + "x-imagesharp" + Path.GetExtension(filePath);
using (FileStream stream = File.OpenRead(filePath))
using (FileStream output = File.OpenWrite(tmpFilePath))
{
	Image image = new Image(stream);
	image.Resize(200, 300).Save(output);
}
sw1.Stop();
Console.WriteLine("Image sharp: " + sw1.ElapsedMilliseconds);
```

### System Configuration
- ImageSharp version: 1.0.0-alpha1-00073
- Other ImageSharp packages and versions: 1.0.0-alpha1-00073 && 1.0.0-alpha1-00053
- Environment (Operating system, version and so on): OSX 10.11.6
- .NET Framework version: Mono / .NET 4.5
- Additional information: Link to the picture (it's the interior of a mouth, so if you don't want to click on it, no problem. Nothing gore or anything though :D): http://i.imgur.com/TSdnmgr.jpg

**Comment by JimBobSquarePants** [comment]
Hmmm... Something amiss here. 946ms on my laptop.

Could you do me a favor and tell me what the results of `Vector.IsHardwareAccelerated` is on your machine? I'm assuming false. 

Mono supports NET4.6 as I recall, would it be possible to bump to that and get back to us?

I'll double check we're not doing something daft and releasing our code in debug mode also.

**Comment by JimBobSquarePants** [comment]
Definitely in release mode. https://github.com/JimBobSquarePants/ImageSharp/blob/f14dd6f91adcda9569bbf49fdff6c613ac2948bd/build/Program.cs#L106

**Comment by dlemstra** [comment]
I wonder which part is slow, what is the output of the following code.

```C#
Stopwatch sw1 = new Stopwatch();
sw1.Start();
string tmpFilePath = Path.GetTempPath() + "x-imagesharp" + Path.GetExtension(filePath);
using (FileStream stream = File.OpenRead(filePath))
using (FileStream output = File.OpenWrite(tmpFilePath))
{
	Image image = new Image(stream);
	sw1.Stop();
	Console.WriteLine("Image sharp read: " + sw1.ElapsedMilliseconds);
	sw1.Restart();
	image.Resize(200, 300);
	sw1.Stop();
	Console.WriteLine("Image sharp resize: " + sw1.ElapsedMilliseconds);
	sw1.Restart();
	image.Save(output);
	sw1.Stop();
	Console.WriteLine("Image sharp save: " + sw1.ElapsedMilliseconds);
}
```

**Comment by antonfirsov** [comment]
@nah0y I'm afraid, there is no hardware acceleration for SIMD types in your environment. After a quick (non-exhausting) googling it seems to me, Mono started to support HW acceleration for `System.Numerics.Vectors` [only a month ago](http://www.mono-project.com/news/2016/12/20/system-numeric-vectors/).

What is the value of [System.Numerics.Vector.IsHardwareAccelerated](https://msdn.microsoft.com/en-us/library/system.numerics.vector.ishardwareaccelerated(v=vs.111).aspx) in release mode?

If it's false, the only way to make ImageSharp faster is to upgrade to the latest mono framework.

**Comment by antonfirsov** [comment]
Renamed the issue to have more information about the context.

**Comment by nah0y** [comment]
Hey guys! Sorry for the wait, I'm back with more info :)

I tried on my new Macbook Pro and same result. 9762ms.
I'm targetting x86.
Vector.IsHardwareAccelerated returns false in release mode
Using .NET 4.5.1: 9686ms.
I don't have .Net 4.6 on this machine for now sorry, I'll have to install it, but not today :/

I don't understand why you wrote this :/

> Definitely in release mode. https://github.com/JimBobSquarePants/ImageSharp/blob/f14dd6f91adcda9569bbf49fdff6c613ac2948bd/build/Program.cs#L106

For info, in the output console I have around 12 threads that are started one after the other:
> Thread started: <Thread Pool> #2
> Thread started: <Thread Pool> #3
> Thread started: <Thread Pool> #4
> Thread started: <Thread Pool> #5
> Thread started: <Thread Pool> #6
> Thread started: <Thread Pool> #7
> Thread started: <Thread Pool> #8
> Thread started: <Thread Pool> #9
> Thread started: <Thread Pool> #10

And from what I understand it's because of Vector.IsHardwareAccelerated is false and I need the latest mono runtime (thanks @antonfirsov for clarifying that). If I don't want to build from sources and wait for an update, it's an update of Mono right? Like currently they're at 4.6, so I'll have to wait for 4.7. I'll try to give it a try when it's available if I don't forget 😄

**Comment by nah0y** [comment]
And @dlemstra, here's the result:

Image sharp read: 5360
Image sharp resize: 4393
Image sharp save: 34

**Comment by antonfirsov** [comment]
@nah0y yeah you need a new mono version, [4.8 should be the next one](http://www.mono-project.com/docs/about-mono/releases/), but I see no info about `System.Numerics.Vectors` in their roadmap. Maybe you should ask on their forums.

Thanks for sharing your results. Looking at resize performance is on my list. Strange it's that slow on Mono.

**Comment by nah0y** [comment]
Yes strange ;)
Thanks a lot to all of you for your time!

**Comment by antonfirsov** [comment]
@nah0y I don't know if you are still interested, but recently I figured out that SIMD support will be shipped [with mono 5.0](http://www.mono-project.com/docs/about-mono/releases/5.0.0/).

We are also working on perf. improvements meanwhile.

**Comment by nah0y** [comment]
@antonfirsov Hey! Thanks for the news. I finally did it another way. But I'm still interested to see how it evolves!

**Comment by clegendre** [comment]
Hello,

I'm experiencing the same issue on Xamarin Andoid version 7.3 (based on Mono 5).
The Image.Load and Image.Save methods are taking so much times on large images (4000x3000 px). 
On a LG G3 mobile running Android 5.0, it takes approximatively 1 minute to Load / Manipulate Exifs / Save in a MemoryStream. 
@antonfirsov I'm interested on how you did it another way. Which library did you used ?

ImageSharp version: 1.0.0-alpha9-00054

**Comment by antonfirsov** [comment]
@clegendre Unfortunately I'm unfamiliar with mobile development and the details of the Xamarin architecture. `4k*3k` means lots of pixels, what are the expected times for that operation (using other libraries/applications)? 

Also: What is the value of  [System.Numerics.Vector.IsHardwareAccelerated](https://msdn.microsoft.com/en-us/library/system.numerics.vector.ishardwareaccelerated(v=vs.111).aspx) on your device? Are you sure you tested it in release mode on your device?

It seems, that it's not necessary to decode the image pixels in your use case, only the exif data + resave the modified exif with the original pixel data. @JimBobSquarePants shouldn't we add support for this?

**Comment by JimBobSquarePants** [comment]
@antonfirsov I believe a standalone metadata editor would be the best tool for that job of which there are several out in the wild. Could we build one? Yeah certainly, Probably pretty easy with the existing tools we have now but trying to configure the existing decoders to only parse metadata while ignoring the pixels seems like it would add a lot of complexity and overhead to our code. 

@clegendre Could you do me a favor and check the value Anton asked for and also post a code sample + device specifications. 1 min is slow... crazy slow even without SIMD. I don't see how we are doing enough work to even cause that much time to be consumed.

**Comment by clegendre** [comment]
_System.Numerics.Vector.IsHardwareAccelerated_ is set to true.
Tested in Debug and Release mode.

We've also tested on a new Lenovo smartphone under Android 7: images taken with the camera are slightly bigger, but even with a better proc, loading then saving in memory took approximatively 20-30 seconds.

@antonfirsov We managed to do it another way, by setting the Exifs directly on the Camera before the image is taken (this is indeed the right way to do that). Thumbnail is done using the native Android API.

@JimBobSquarePants I'll try to post a repro sample this weekend. 

BTW, I love your library and the API design is fantastic, you're making a great job!
